src/machine.rs

```diff
@@ -62,57 +62,69 @@ pub struct Frame<'tcx> {
     /// return to any particular address.
     pub ret_block: Option<BasicBlock>,
 
-    /// We make this optional for the purpose of optimisation. Some functions do
-    /// not need a locals field (e.g some closures and 0-arg functions).
-    /// Functions that do require a local env can have it explicitly
-    /// instantiated lazily when required.
-    locals: IndexVec<Local, Vec<u8>>
+    /// This field serves as a lookup to a vars position in the contiguous
+    /// `locals` byte vector, where the index is the local id and the value is
+    /// the corresponding index into the `locals` field where the value's bytes
+    /// start.
+    ///
+    /// This should not be accessed directly.
+    idxs: Vec<usize>,
+
+    /// An exact sized byte vector containing the frame local vars in their byte
+    /// representation. These values can change at runtime and this field should
+    /// not be accessed directly.
+    locals: Vec<u8>,
 }
 
 impl<'tcx> Frame<'tcx> {
     pub fn new(def_id: DefId,
                mir: &'tcx Mir<'tcx>,
-               locals: IndexVec<Local, Vec<u8>>,
+               idxs: Vec<usize>,
+               locals: Vec<u8>,
                ret_val: Option<Address>,
                ret_block: Option<BasicBlock>)
                -> Self {
         Frame {
             def_id: def_id,
             mir: mir,
+            idxs: idxs,
             locals: locals,
             ret_val: ret_val,
             ret_block: ret_block,
         }
     }
 
-    fn set_local(&mut self, local: Local, bytes: Vec<u8>) {
-        self.locals[local] = bytes
-    }
-
-    fn get_local(&self, local: Local) -> &[u8] {
-        self.locals[local].as_slice()
+    pub fn local_ptr(&self, local: Local) -> Ptr {
```

Comment:
So this gets the Ptr for a local variable from the aggregate local storage?

Why do you have to special case if it's the end?

Reply:
Because in addition to the offset, the `Ptr` takes a `size` argument to know how many bytes to read / write to. 

It does this in a somewhat hacky way. It gets the index to the next local and uses that to determine the size. Hence, the last item in the `Vec` will be an edge case.